hyperspy/tests/utils/test_roi.py

```diff
@@ -486,6 +486,37 @@ def test_line2droi_angle(self):
         with pytest.raises(ValueError):
             r.angle(axis='z')
 
+    def test_repr_None(self):
+        # Setting the args=None sets them as traits.Undefined, which didn't 
+        # have a string representation in the old %s style.
+        repr(Point1DROI())
+        repr(Point2DROI())
+        repr(Line2DROI())
+        repr(RectangularROI())
+        repr(SpanROI())
+        repr(CircleROI())
+
+    def test_repr_vals(self):
+        repr(Point1DROI(1.1))
+        repr(Point2DROI(1.1, 2.1))
+        repr(Line2DROI(0, 0, 1, 1, 0.1))
+        repr(RectangularROI(0, 0, 1, 1))
+        repr(SpanROI(3., 5.))
+        repr(CircleROI(5, 5, 3))
+        repr(CircleROI(5, 5, 3, 1))
+
+    def test_undefined_call(self):
+        rect = RectangularROI(None, None, None, None)
+        with pytest.raises(AttributeError, match='not yet been set'):
+            rect(self.s_s)
+
+    def test_undefined_call_circle(self):
+        circ1 = CircleROI(None, None, None, None)
+        with pytest.raises(AttributeError, match='not yet been set'):
+            circ1(self.s_s)
```

Comment:
Do this test for all ROI, I would add a test which create the ROI then slice a signal to test `__call__`. This test is very fast and it is good to make sure that all ROI follows a well defined and consistent behaviour to avoid the current issue we have with `RectangularROI` for example!

Reply:
So far I have gone for the following, which is what I interpreted as what you meant, but I'm not sure how that is supposed to deal with the ordering issue with RectangularROI.

```python
    def test_undefined_call(self):
        for roi in [Point1DROI, Point2DROI, RectangularROI, SpanROI, Line2DROI, CircleROI]:
            r = roi()
            with pytest.raises(ValueError, match='not yet been set'):
                r(self.s_s)

    def test_default_values_call(self):
        for roi in [Point1DROI, Point2DROI, RectangularROI, SpanROI, Line2DROI, CircleROI]:
            r = roi()
            r._set_default_values(self.s_s)
            r(self.s_s)
```

Reply:
What needs to be tested here is that all ROI set the parameters after `_set_default_values`. If it doesn't fit well in a for loop, maybe it is better to do one ROI the one after another. I would keep your test as it is and add another test using interactive and checking that the values are correctly set by `_set_default_values`. These different tests will test slightly different supported usages.